### PR TITLE
Feat: Add Python Matic Exit script

### PR DESCRIPTION
**Pull request by skellet0r**
This was a doooozzzyyyy, but it adds a fresh Python script `exit.py` for withdrawing an ERC20 asset from matic.
The script handles both burning the asset on matic, checking for burn tx checkpoint inclusion on ethereum, and withdrawing the asset back into your acct.

Good to merge :+1: 

Testing can be done by running `brownie run exit tester --network mainnet`

Depends on: #13

### CODE REVIEW

File: scripts/exit.py [hunk 1]
```diff
@@ -4,19 +4,435 @@
...
+PreparedReceipt = Tuple[bytes, int, bytes, PreparedLogs]
+
 
 # MUST SET VARIABLES BEFORE BURNING ON MATIC
 MSG_SENDER = accounts.add()
 MATIC_ERC20_ASSET_ADDR = ""
 BURN_AMOUNT = 0
 
+# BURN TX HASH
+MATIC_BURN_TX_ID = "0x4486e398e0f2ca4d00bec85edbb9aff94e7085fa2b5ef18319989d9d8e37152f"
```
Comment (iamdefinitelyahuman) [inline scripts/exit.py]: Should probably leave this blank by default.

File: scripts/exit.py [hunk 2]
```diff
@@ -4,19 +4,435 @@
...
+        rlp.encode(receipt_proof),
+        HexBytes(0) + path,
+        log_index,
+    ]
+    return rlp.encode(payload)
+
+
+def build_calldata(burn_tx_id: str = MATIC_BURN_TX_ID) -> bytes:
+    """Generate the calldata required for withdrawing ERC20 asset on Ethereum."""
+    assert is_burn_checkpointed()
```
Comment (iamdefinitelyahuman) [inline scripts/exit.py]:
```suggestion
    assert is_burn_checkpointed(burn_tx_id)
```

File: scripts/exit.py [hunk 3]
```diff
@@ -4,19 +4,435 @@
...
+    calldata = build_calldata()
+
+    deployment_addrs = fetch_deployment_data()["Main"]["POSContracts"]
+    root_chain_mgr_proxy_addr, root_chain_mgr_addr = (
+        deployment_addrs["RootChainManagerProxy"],
+        deployment_addrs["RootChainManager"],
+    )
+    root_chain_mgr = Contract.from_explorer(
+        root_chain_mgr_proxy_addr, root_chain_mgr_addr, silent=True
+    )
```
Comment (iamdefinitelyahuman) [inline scripts/exit.py]:
```suggestion
    root_chain_mgr = Contract(root_chain_mgr_proxy_addr)
```

Adding `autofetch_sources: True` to the config file will allow us to do away with `from_explorer` and speed things up due to caching.  Also, when testing I ran into an issue in that matic's json providing the addresses points at an out-of-date implementation contract.  So better to just dynamically determine, I think.

File: scripts/exit.py [hunk 4]
```diff
@@ -4,19 +4,435 @@
...
+        burn_tx_block["receiptsRoot"],
+        burn_tx_receipt,
+        receipt_proof,
+        path,
+        log_index,
+    )
+    return calldata
+
+
+def exit():
```
Comment (iamdefinitelyahuman) [inline scripts/exit.py]: `exit` is a function in python, maybe better to use a different name?

File: scripts/exit.py [hunk 5]
```diff
@@ -4,19 +4,435 @@
...
+    receipts_root = HexBytes(block["receiptsRoot"])
+    return keccak256(block_number + timestamp + txs_root + receipts_root)
+
+
+@hot_swap_network("polygon-main")
+def build_block_proof(block_start: int, block_end: int, burn_tx_block_number: int) -> List[bytes]:
+    """Build a merkle proof for the burn tx block."""
+    checkpoint_blocks = (
+        web3.eth.get_block(block_number) for block_number in range(block_start, block_end + 1)
+    )
```
Comment (iamdefinitelyahuman) [inline scripts/exit.py]: 1. You could use brownie's `chain` object here
2. This is slooow and at first I thought the script was hanging. Some console output to appease impatient users would be pretty lovely.

File: scripts/exit.py (hunk 4 above)
Comment (iamdefinitelyahuman) [reply scripts/exit.py]: It also took me a bit to figure out that this was the function I want - there's a lot going on inside this script! Providing a `main` method would be nice.

File: scripts/exit.py [hunk 6]
```diff
@@ -4,19 +4,435 @@
...
+def build_calldata(burn_tx_id: str = MATIC_BURN_TX_ID) -> bytes:
+    """Generate the calldata required for withdrawing ERC20 asset on Ethereum."""
+    assert is_burn_checkpointed()
+
+    burn_tx, burn_tx_receipt, burn_tx_block = fetch_burn_tx_data(burn_tx_id)
+    start, end, header_block_number = fetch_block_inclusion_data(burn_tx_block["number"])
+    block_proof = build_block_proof(start, end, burn_tx_block["number"])
+    path, receipt_proof = build_receipt_proof(burn_tx_receipt, burn_tx_block)
+
+    log_index = find_log_index(burn_tx_receipt)
```
Comment (iamdefinitelyahuman) [inline scripts/exit.py]: In the event of an incorrect txid, this function will raise. If we move it to before the proof generation, we can save ourselves a few minutes when there's a mistake.

File: scripts/exit.py [hunk 7]
```diff
@@ -4,19 +4,435 @@
...
 def burn_asset_on_matic():
+    """Burn an ERC20 asset on Matic Network"""
     ChildERC20 = load_project("maticnetwork/pos-portal@1.5.2").ChildERC20
     asset = ChildERC20.at(MATIC_ERC20_ASSET_ADDR)
     tx = asset.withdraw(BURN_AMOUNT, {"from": MSG_SENDER})
     print("Burn transaction has been sent.")
     print(f"Visit https://explorer-mainnet.maticvigil.com/tx/{tx.txid} for confirmation")
+
+
+def hot_swap_network(network_name):
```
Comment (iamdefinitelyahuman) [inline scripts/exit.py]: This is very cool :) Could it be tweaked some to allow using the script on the testnet as easy as it is on mainnet?

Brownie should offer first-class support for this sort of behavior :thinking:

**Review by iamdefinitelyahuman — COMMENTED** [review]
Very cool, just a few tweaks I'd like to see :pray:

**Comment by skellet0r** [comment]
It feels weird making that main function entrypoint with `input`... wait could I have used `click` .... :thinking: 

Edit: Too much complexity already, no need for more

**Comment by skellet0r** [comment]
This script is sooo bulky, I feel like eventually I should take out the matic portion in make it a seperate library :/
But now all one has to do to call it is:
brownie run exit --network ??
It should work with testnets too ... let me check that :eyes: 


Edit: yeh, it works

**Comment by skellet0r** [comment]
I added some progress info for the checkpointing because I got tired of running the script

**Review by iamdefinitelyahuman — APPROVED** [review]

**Pull request merged**